modules/channel-messenger/src/backend/messenger.ts

```diff
@@ -0,0 +1,253 @@
+import { Router } from 'express'
+import _ from 'lodash'
+
+import { Config } from '../config'
+
+const outgoingTypes = ['text', 'typing', 'login_prompt', 'carousel']
+type MessengerAction = 'typing_on' | 'typing_off' | 'mark_seen'
+
+export class MessengerService {
+  private readonly http = axios.create({ baseURL: 'https://graph.facebook.com/v2.6/me' })
+
+  private messengerClients: { [key: string]: MessengerClient } = {}
+  private router: Router
+
+  constructor(private bp: typeof sdk) {}
+
+  initialize() {
+    this.router = this.bp.http.createRouterForBot('channel-messenger', { checkAuthentication: false })
+    this.router.use(
+      bodyParser.json({
+        verify: this._verifySignature.bind(this)
+      })
+    )
+
+    this.router.get('/webhook', this._setupWebhook.bind(this))
+    this.router.post('/webhook', this._handleIncomingMessage.bind(this))
+
+    this.bp.events.registerMiddleware({
+      description: 'Sends outgoing messages for the messenger channel',
+      direction: 'outgoing',
+      handler: this._handleOutgoingEvent.bind(this),
+      name: 'messenger.sendMessages',
+      order: 200
+    })
+  }
+
+  getMessengerClient(botId: string): MessengerClient {
+    if (this.messengerClients[botId]) {
+      return this.messengerClients[botId]
+    }
+
+    this.messengerClients[botId] = new MessengerClient(botId, this.bp, this.http)
+    return this.messengerClients[botId]
+  }
+
+  // See: https://developers.facebook.com/docs/messenger-platform/webhook#security
+  private async _verifySignature(req, res, buffer) {
+    const client = this.getMessengerClient(req.params.botId)
+    const config = await client.getConfig()
+    const signatureError = new Error("Couldn't validate the request signature.")
+
+    if (!/^\/webhook/i.test(req.path)) {
+      return
+    }
+
+    const signature = req.headers['x-hub-signature']
+    if (!signature) {
+      throw signatureError
+    } else {
+      const [, hash] = signature.split('=')
+      const expectedHash = crypto
+        .createHmac('sha1', config.verifyToken)
+        .update(buffer)
+        .digest('hex')
+
+      if (hash != expectedHash) {
+        throw signatureError
+      }
+    }
+  }
+
+  private async _handleIncomingMessage(req, res) {
+    const body = req.body
+    const botId = req.params.botId
+    const client = this.getMessengerClient(botId)
+
+    if (body.object !== 'page') {
+      res.sendStatus(404)
+      return
+    }
+
+    for (const entry of body.entry) {
+      // Will only ever contain one message, so we get index 0
+      const webhookEvent = entry.messaging[0]
+      const senderId = webhookEvent.sender.id
+
+      await client.sendAction(senderId, 'mark_seen')
+
+      if (webhookEvent.message) {
+        await this._sendEvent(botId, senderId, webhookEvent.message, { type: 'message' })
+      } else if (webhookEvent.postback) {
+        await this._sendEvent(botId, senderId, { text: webhookEvent.postback.payload }, { type: 'callback' })
```

Comment:
In my world postback != callback. A callback is a function and a postback a payload that is __posted back__ to your app / bot